validator.go

```diff
@@ -1,16 +1,125 @@
 package egcl
 
 import (
+	"fmt"
 	egdm "github.com/mimiro-io/entity-graph-data-model"
+	"github.com/pkg/errors"
+	"math"
 )
 
+type DataProvider interface {
+	Hop(sourceEntityId string, reference string, inverse bool) ([]*egdm.Entity, error)
+	GetEntity(entityId string) (*egdm.Entity, error)
+	GetDatasetEntities(name string) ([]*egdm.Entity, error)
+}
+
+type SchemaValidator interface {
+	ValidateEntity(schema *Schema, entity *egdm.Entity) (ok bool, exceptions []*ConstraintViolation, err error)
+	ValidateDataset(schema *Schema, datasetName string) (ok bool, exceptions []*ConstraintViolation, err error)
+	ValidateSchema(schema *Schema) (ok bool, exceptions []*ConstraintViolation, err error)
+	ValidateEntityCollection(schema *Schema, entityCollection *egdm.EntityCollection) (ok bool, exceptions []*ConstraintViolation, err error)
+}
+
+type NamespaceResolver interface {
+	ResolveNamespace(prefix string) (string, error)
+}
+
+type ValidatorSettings struct {
+	// StrictValidation will cause the validator to fail if an entity has properties or references that are not defined in the schema
+	StrictValidation bool
+	// ValidateRelated enables validation of referenced entities, specifically that they exist and are of the correct type
+	ValidateRelated bool
+	// datasetsContext defines the set of datasets that are in play when checking referenced entities
+	DatasetsContext []string
+}
+
 type Validator struct {
+	settings     *ValidatorSettings
+	dataProvider DataProvider
+}
+
+func NewValidator() *Validator {
+	return &Validator{}
+}
+
+func (v *Validator) WithSettings(settings *ValidatorSettings) *Validator {
+	v.settings = settings
+	return v
 }
 
+func (v *Validator) WithDataProvider(provider DataProvider) *Validator {
+	v.dataProvider = provider
+	return v
+}
+
+type ViolationType int
+
+const (
+	MinPropertyOccurrenceNotMet ViolationType = iota
+	MaxPropertyOccurrenceExceeded
+	MinReferenceOccurrenceNotMet
+	MaxReferenceOccurrenceExceeded
+	ReferenceNotFound
+	ReferenceTypeMismatch
+	AbstractEntityClassViolation
+)
+
+// for our purposes, we can use math.MaxInt32 as infinity
+const MaxInt = math.MaxInt32
+
 type ConstraintViolation struct {
+	Constraint    any
+	Entity        *egdm.Entity
+	ViolationType ViolationType
+	Message       string
 }
 
-func (v *Validator) Validate(schema *Schema, entity *egdm.Entity) (ok bool, exceptions []*ConstraintViolation, err error) {
+func NewConstraintViolation(constraint any, entity *egdm.Entity, violationType ViolationType, message string) *ConstraintViolation {
+	violation := &ConstraintViolation{}
+	violation.Constraint = constraint
+	violation.Entity = entity
+	violation.ViolationType = violationType
+	violation.Message = message
+	return violation
+}
+
+// ValidateDataset validates the given dataset against the data in the named dataset
+func (v *Validator) ValidateDataset(schema *Schema, datasetName string) (ok bool, exceptions []*ConstraintViolation, err error) {
```

Comment:
Having both bool and error returned can be confusing. Eg what does it mean if ok=true and err!=nil

Reply:
decided to keep this as the errors are different from validation exceptions. the ok is a shorthand for checking the exceptions is nil. Question is valid, but not different to what does it mean if exceptions is not empty and there are errors.  As with all go, errors should take precedence if non-nil over other func return values.